src/NonCumulative/Soundness/Realizability.agda

```diff
@@ -0,0 +1,527 @@
+  with presup-s-≈ σ≈
+... | _ , ⊢σ , _ , _
+    with presup-s (⊢w⇒⊢s ⊢τ)
+... | _ , ⊢∷ ⊢Γ ⊢T′ = begin
+  v (len Γ₁) [ σ ] ≈⟨ []-cong (v-≈ ⊢Γ n∈) σ≈ ⟩
+  v (len Γ₁) [ p τ ] ≈⟨ ≈-conv ([∘] ⊢τ′ (s-wk ⊢TΓ) (vlookup ⊢Γ n∈)) (≈-sym [wkτ]≈) ⟩
+  v (len Γ₁) [ wk ] [ τ ] ≈⟨ ≈-conv ([]-cong ([wk] ⊢Γ ⊢T′ n∈) (s-≈-refl ⊢τ′)) wkτ≈ ⟩
+  v (1 + len Γ₁) [ τ ] ≈⟨ ≈-conv (v0∼x-gen (( T′ ↙ i₁) ∷ Γ₁) ⊢τ refl) wkτ≈ ⟩
+  v (len Δ ∸ len Γ₂ ∸ 1) ∎
+  where open ER
+        n∈      = n∶T[wk]n∈!ΔTΓ Γ₁ refl
+        ⊢TΓ     = ⊢∷ ⊢Γ ⊢T′
+        ⊢τ′     = ⊢w⇒⊢s ⊢τ
+        ⊢T[wk]* = proj₂ (presup-tm (⊢vn∶T[wk]suc[n] ⊢Γ refl))
+        [wkτ]≈  = []-cong-Se″ ⊢T[wk]* ⊢σ σ≈
+        wkτ≈    = ≈-trans ([∘]-Se ⊢T[wk]* (s-wk ⊢TΓ) ⊢τ′) (≈-sym [wkτ]≈)
+
+v0∼x : ∀ {i} (A≈B : A ≈ B ∈ 𝕌 i) →
+       Γ ⊢ T ®[ i ] A≈B →
+       (T ↙ i) ∷ Γ ⊢ v 0 ∶ T [ wk ] ®↓[ i ] l (len Γ) ∈El A≈B
+v0∼x {_} {_} {Γ} A≈B T∼A
+  with ®⇒ty A≈B T∼A
+...  | ⊢T
+     with presup-tm ⊢T
+...     | ⊢Γ , _ = record
+  { t∶T  = vlookup ⊢TΓ here
+  ; T∼A  = ®-one-sided A≈B A≈B (®-mon′ A≈B T∼A (r-p (⊢wI ⊢TΓ) (s-≈-sym (∘-I (s-wk ⊢TΓ)))))
+  ; c∈⊥  = Bot-l (len Γ)
+  ; krip = λ {Δ} {σ} ⊢σ → v0∼x-gen [] ⊢σ refl
+  }
+  where ⊢TΓ = ⊢∷ ⊢Γ ⊢T
+
+private
+  module Real where
+    mutual
+      ®↓El⇒®El : ∀ {i} → (rc : ∀ {j} → j < i → ∀ {A B Γ T Δ σ} (A≈B : A ≈ B ∈ 𝕌 j) → Γ ⊢ T ®[ j ] A≈B → Δ ⊢w σ ∶ Γ → ∃ λ W → Rty len Δ - A at j ↘ W × Δ ⊢ T [ σ ] ≈ Nf⇒Exp W ∶[ 1 + j ] Se j) →
+                 (A≈B : A ≈ B ∈ 𝕌 i) → Γ ⊢ t ∶ T ®↓[ i ] c ∈El A≈B → Γ ⊢ t ∶ T ®[ i ] ↑ i A c ∈El A≈B
+      ®↓El⇒®El rc R@(ne′ C≈C′) t®↓ = (ne c∈⊥ refl refl) , record
+        { t∶T = t∶T
+        ; ⊢T = ®⇒ty R T∼A
+        ; krip = λ ⊢σ → proj₂ T∼A ⊢σ , (krip ⊢σ)
+        }
+        where open _⊢_∶_®↓[_]_∈El_ t®↓
+      ®↓El⇒®El rc N′ t®↓ = ne c∈⊥ (λ ⊢σ → ≈-conv (krip ⊢σ) (≈-trans ([]-cong-Se′ T∼A (⊢w⇒⊢s ⊢σ)) (N-[] (⊢w⇒⊢s ⊢σ)))) , T∼A
+        where open _⊢_∶_®↓[_]_∈El_ t®↓
+      ®↓El⇒®El rc U′ t®↓ = record
+        { t∶T = t∶T
+        ; T≈ = T∼A
+        ; A∈𝕌 = ne c∈⊥ refl refl
+        ; rel = (conv t∶T T∼A) , λ ⊢σ → ≈-conv (krip ⊢σ) (≈-trans ([]-cong-Se′ T∼A (⊢w⇒⊢s ⊢σ)) (Se-[] _ (⊢w⇒⊢s ⊢σ)))
+        }
+        where open _⊢_∶_®↓[_]_∈El_ t®↓
+      ®↓El⇒®El {Π j A (S ↙ k) ρ} {_} {Γ} {t} {T} {c} rc (Π′ jA RT) t®↓ = record
+        { t∶T = t∶T
+        ; a∈El = El-refl (Π′ jA RT) (Bot⊆El (Π′ jA RT) c∈⊥)
+        ; IT = IT
+        ; OT = OT
+        ; ⊢IT = ⊢IT
+        ; ⊢OT = ⊢OT
+        ; T≈ = T≈
+        ; krip = λ {Δ} {σ} ⊢σ → record
+          { IT-rel = ΠRel.IT-rel (G.krip ⊢σ)
+          ; ap-rel = λ s® b∈ →
+            let a , ↘a , ∼a = helper refl ⊢σ jA RT b∈ s® G.krip in
+            record
+            { fa = a
+            ; ↘fa = ↘a
+            ; ®fa = ∼a
+            }
+          }
+        }
+        where module ↓ = _⊢_∶_®↓[_]_∈El_ t®↓
+              open ↓
+              module G = GluΠ T∼A
+              open G
+              helper : ∀ {k′} →
```

Comment:
can we make use of the `inv` lemmas and make it a bit nicer? we can discuss

Reply:
I think, to achieve so, we need to generalize the defs of `_⊢_∶_®[_]_∈El_`, `_⊢_∶_®↓[_]_∈El_`and `_⊢_∶_®↑[_]_∈El_`, to make them irrelevant to the concrete proof of `A ≈ B ∈ 𝕌 i`. we can discuss this in detail in the meeting. since it's the only place requiring this, not sure if it's worthwhile to do so